src/main/java/datawave/microservice/authorization/user/ProxiedUserDetails.java

```diff
@@ -24,33 +23,69 @@
-     * {@link DatawaveUser#getUserType()} is {@link UserType#USER}. If no proxied principal is a user principal, then the first principal in the list is
-     * returned.
+     * Gets the {@link DatawaveUser} that represents the primary user in this ProxiedUserDetails. If there is only one DatawaveUser, then it is the primaryUser.
+     * If there is more than one DatawaveUser, then the first (and presumably only) DatawaveUser whose {@link DatawaveUser#getUserType()} is
+     * {@link UserType#USER} is the primary user. If no such DatawaveUser is present, then the second principal in the list is returned as the primary user.
+     * This will be the first entity in the X-ProxiedEntitiesChain which should be the server that originated the request.
      *
-     * @return The {@link DatawaveUser} that represents the user in the list of proxied users
+     * @return The {@link DatawaveUser} that represents the primary user in the list of proxied users
      */
     @JsonIgnore
     public DatawaveUser getPrimaryUser() {
-        DatawaveUser first = proxiedUsers.stream().findFirst().orElse(null);
-        return proxiedUsers.stream().filter(u -> u.getUserType() == UserType.USER).findFirst().orElse(first);
+        return ProxiedUserDetails.findPrimaryUser(this.proxiedUsers);
+    }
+    
+    static protected DatawaveUser findPrimaryUser(List<DatawaveUser> datawaveUsers) {
+        if (datawaveUsers.size() <= 1) {
+            return datawaveUsers.get(0);
+        } else {
+            DatawaveUser secondInOrder = datawaveUsers.get(1);
+            return datawaveUsers.stream().filter(u -> u.getUserType() == UserType.USER).findFirst().orElse(secondInOrder);
+        }
+    }
+    
+    /*
+     * The purpose here is to return a List of DatawaveUsers where the original caller is first followed by any entities in X-ProxiedEntitiesChain in the order
+     * that they were traversed and ending with the entity that made the final call. The List that is passed is not modified. This method makes the following
+     * assumptions about the List that is passed to ths method: 1) The first element is the one that made the final call 2) Additional elements (if any) are
+     * from X-ProxiedEntitiesChain in chronological order of the calls
+     */
+    static protected List<DatawaveUser> orderProxiedUsers(List<DatawaveUser> datawaveUsers) {
+        DatawaveUser primary = ProxiedUserDetails.findPrimaryUser(datawaveUsers);
+        List<DatawaveUser> users = new ArrayList<>();
+        users.add(primary);
+        if (datawaveUsers.size() > 1) {
+            // @formatter:off
+            // Skipping first user because if it is UserType.USER, it is the primary
+            // and already added.  If it is UserType.Server, then it will be added at the end
+            users.addAll(datawaveUsers.stream()
+                    .skip(1)
+                    .filter(u -> u != primary)
+                    .collect(Collectors.toList()));
```

Comment:
Brian used to add this comment for me all the time, so I'm going to add it for you...

Consider using forEach vs. creating a new list.

i.e.

```
datawaveUsers.stream()
   .skip(1)
   .filter(u -> u != primary)
   .forEach(x -> users.add(x));
```

Reply:
@jwomeara - made changes